packages/migrations/src/batched-migrations/batched-migrations-runner.ts

```diff
@@ -0,0 +1,318 @@
+   * an underscore with additional characters, which are ignored. These should
+   * typically be used to provide a human-readable name for the migration.
+   */
+  private async loadMigrationImplementation(migrationFile: MigrationFile) {
+    // We use dynamic imports to handle both CJS and ESM modules.
+    const migrationModulePath = path.join(migrationFile.directory, migrationFile.filename);
+    const migrationModule = await import(migrationModulePath);
+
+    const migrationImplementation = migrationModule.default as BatchedMigrationImplementation;
+    validateBatchedMigrationImplementation(migrationImplementation);
+    return migrationImplementation;
+  }
+
+  async enqueueBatchedMigration(identifier: string) {
+    const migrationFile = await this.getMigrationForIdentifier(identifier);
+    if (!migrationFile) {
+      throw new Error(`No migration found for identifier ${identifier}`);
+    }
+
+    const migrationImplementation = await this.loadMigrationImplementation(migrationFile);
+    const migrationParameters = await migrationImplementation.getParameters();
+
+    // If `max` is null, that implies that there are no rows to process, so
+    // we can immediately mark the migration as finished.
+    const status: BatchedMigrationStatus =
+      migrationParameters.max === null ? 'succeeded' : 'pending';
+
+    const minValue = migrationParameters.min ?? DEFAULT_MIN_VALUE;
+    const maxValue = migrationParameters.max ?? minValue;
+    const batchSize = migrationParameters.batchSize ?? DEFAULT_BATCH_SIZE;
+
+    await insertBatchedMigration({
+      project: this.options.project,
+      filename: migrationFile.filename,
+      timestamp: migrationFile.timestamp,
+      batch_size: batchSize,
+      min_value: minValue,
+      max_value: maxValue,
+      status,
+    });
+  }
+
+  async finalizeBatchedMigration(identifier: string) {
+    const timestamp = identifier.split('_')[0];
+
+    let migration = await selectBatchedMigrationForTimestamp(this.options.project, timestamp);
+
+    if (migration.status === 'succeeded') return;
+
+    // If the migration isn't already in the finalizing state, mark it as such.
+    if (migration.status !== 'finalizing') {
+      migration = await updateBatchedMigrationStatus(migration.id, 'finalizing');
+    }
+
+    await doWithLock(this.lockNameForTimestamp(timestamp), { autoRenew: true }, async () => {
+      const migrationFile = await this.getMigrationForIdentifier(identifier);
+      if (!migrationFile) {
+        throw new Error(`No migration found for identifier ${identifier}`);
+      }
+      const migrationImplementation = await this.loadMigrationImplementation(migrationFile);
+
+      const runner = new BatchedMigrationRunner(migration, migrationImplementation, {
+        logProgress: true,
+      });
+
+      // Because we don't give any arguments to `run()`, it will run until it
+      // has attempted every job.
+      await runner.run();
+    });
+
```

Comment:
Oh actually I guess there is still no risk of overloading the database since its all being done in batches still anyway. 

I still don't see the point of forcing a single machine to synchronously do all the jobs though.

Reply:
Automatically finishing them is extremely useful for anyone doing local development or self-hosting a single-server installation. If we *didn't* do that, you'd have to very carefully plan a list of revisions to slowly step through, which is highly undesirable. Introducing finalization allows us to maintain the very nice property that one can generally just do an upgrade via `git pull && make start` and have a functioning server with zero intervention.

Reply:
Ok this makes sense. I think if the local dev was the only use case for it, there would be a simpler solution, but I agree this seems very desirable for running with a single instance in production set up.